Improve default DNS

Akariln opened the pull request:
~~Remote proxy server may not support UDP relay,
use TCP to make DNS via proxy;
It is also possible to set up a backup DNS.~~

**Testing found that many public DNS Servers have serious problems with DNS over TCP.
Decided not to change the DNS and simply add backup DNS.**

Mainland DNS usually chooses Direct Outbound,
Consider whether to use the following instead to reduce latency? (bypass Routing)

```JSON
    {
      "address": "tcp+local://223.5.5.5",
      "skipFallback": true,
      "domains": [
        "geosite:cn"
      ],
      "expectIPs": [
        "geoip:cn"
      ]
    }
```

Akariln committed:
Improve default DNS

Proxy may not support UDP relay,
use TCP to make DNS via proxy;
It is also possible to set up a backup DNS.

Akariln commented:
I found that user-defined Mainland DNS is almost never used when "Routing domainStrategy" and "Freedom domainStrategy" are both set to "AsIs" by default.

Personally recommend setting "Freedom domainStrategy" to "UseIPv4" or "UseIPv4v6". 
("UseIP" has a bug where it will sometimes try to connect to an IPv6 address from DNS resolution results even if you don't have an IPv6 address.)

CharlesWou commented:
>VLESS、Trojan、VMess、Mux 都是 UDP over TCP，且走底层传输方式
>https://github.com/XTLS/Xray-core/discussions/237

Proxy(Xray) will encapsulate the UDP packet in TCP packet  and forward it to the server, so proxy(Xray) supports forwarding DNS(UDP 53) packets to the server

>https://xtls.github.io/en/config/dns.html#dnsobject

And Xray official configuration introduction page uses UDP 53 DNS.

CharlesWou commented:
>"AsIs": Use only the domain name for routing selection. Default value.
>https://xtls.github.io/en/config/routing.html#routingobject

"AsIs" does not resolve the domain name to an IP, it directly sends the request to the server in the form of a domain name. After the server resolves the domain name locally and obtains the web page data, it returns web page data directly to the client.

CharlesWou commented:
>When the value is in the form of "tcp+local://host:port", such as "tcp+local://8.8.8.8:53", Xray will use TCP local mode (TCPL) for queries. That is, DNS requests will not pass through the routing component and will directly request outbound through Freedom, to reduce latency. When the port is not specified, the default port 53 is used.
>https://xtls.github.io/en/config/dns.html#dnsobject

I think what he meant was to reduce CPU processing time, but when data packets are transmitted over the network, TCP has much more latency than UDP.

CharlesWou commented:
> > > VLESS、Trojan、VMess、Mux 都是 UDP over TCP，且走底层传输方式
> > > [XTLS/Xray-core#237](https://github.com/XTLS/Xray-core/discussions/237)
> > 
> > 
> > Proxy（Xray）will encapsulate the UDP packet in TCP packet and forward it to the server, so proxy(Xray) supports forwarding DNS(UDP 53) packets to the server
> 
> There might be a misunderstanding. I mean the remote proxy server, not Xray itself.

This repo is a proxy client, the configurations are about the client.
This client will encapsulate the UDP packet in TCP packet and forward it to the server.

CharlesWou commented:
> 当使用 "AsIs" 时，Xray 将直接使用 golang 默认的连接优先级。出于一些原因，UDP连接如果使用域名会无视系统设置优先IPv4。
> https://xtls.github.io/config/outbounds/freedom.html#outboundconfigurationobject

> "AsIs": Freedom resolves the domain name using the system DNS server and connects to it.
> https://xtls.github.io/en/config/outbounds/freedom.html#outboundconfigurationobject


> I found that Mainland DNS is almost never used when "Routing domainStrategy" and "Freedom domainStrategy" are both set to "AsIs" by default.

How did you judge, capture the packet or check the log?

Akariln committed:
Update DNS

Testing found that many public DNS over TCP have serious problems and are not practical.

Akariln committed:
Update

Akariln commented:
Since DNS over TCP is not applicable, this PR is closed.
